### PR TITLE
Replace linears in DeepSeek V3 with nvFuser's TP linears. 

### PR DESCRIPTION
**Pull request by wujingyue**
cc @kshitij12345

### CODE REVIEW

File: tests/python/multidevice/linear.py [hunk 1]
```diff
@@ -3,25 +3,31 @@
...
+from functools import partial, lru_cache
 from fusion_definition_wrapper import FusionDefinitionWrapper
 from nvfuser import DataType, FusionDefinition
 from torch.distributed.tensor import DTensor
+from torch.distributed.tensor.placement_types import Placement
+from typing import Iterable
 
 
-@dataclass
+@dataclass(frozen=True)
```
Comment (wujingyue) [inline tests/python/multidevice/linear.py:17]: So it's hashable and can be used with lru_cache.

File: tests/python/multidevice/test_deepseek_v3.py [hunk 2]
```diff
@@ -62,12 +69,77 @@ def load_model(config: transformers.PretrainedConfig) -> transformers.PreTrained
...
+    assert isinstance(linear, torch.nn.Linear), f"Unsupported layer: {linear}"
+
+    assert len(parallel_style.input_layouts) == 1, "Expect 1D mesh"
+    input_layout = parallel_style.input_layouts[0]
+
+    assert len(parallel_style.output_layouts) == 1, "Expect 1D mesh"
+    output_layout = parallel_style.output_layouts[0]
+
+    if isinstance(parallel_style, RowwiseParallel):
+        assert input_layout.is_shard(-1), "We only implemented TP."
```
Comment (Priya2698) [inline tests/python/multidevice/test_deepseek_v3.py]:
```Suggestion
        assert input_layout.is_shard(-1), "We only implemented TP, which expects inputs to be sharded on the contracting dimension."
```

File: tests/python/multidevice/test_deepseek_v3.py [hunk 3]
```diff
@@ -62,12 +69,77 @@ def load_model(config: transformers.PretrainedConfig) -> transformers.PreTrained
...
+        assert output_layout.is_replicate(), "We only implemented TP."
```
Comment (Priya2698) [inline tests/python/multidevice/test_deepseek_v3.py]:
```Suggestion
        assert output_layout.is_replicate(), "We only implemented TP, which expects outputs to be replicated."

Comment (Priya2698) [reply tests/python/multidevice/test_deepseek_v3.py]: Same for the other error messages.

File: tests/python/multidevice/test_deepseek_v3.py [hunk 4]
```diff
@@ -62,12 +69,77 @@ def load_model(config: transformers.PretrainedConfig) -> transformers.PreTrained
...
+        assert output_layout.is_shard(-1), "We only implemented TP."
+        return TensorParallelLinear.distribute(linear, mesh, [input_layout], [Shard(0)])
+
+    assert False, f"Unsupported parallel style: {parallel_style}"
+
+
+def parallelize_module_with_nvfuser(
+    module: torch.nn.Module,
+    mesh: dist.device_mesh.DeviceMesh,
+    parallel_plan: dict[str, ParallelStyle],
```
Comment (Priya2698) [inline tests/python/multidevice/test_deepseek_v3.py:119]: Code comment the available values.

File: tests/python/multidevice/test_deepseek_v3.py [hunk 5]
```diff
@@ -62,12 +69,77 @@ def load_model(config: transformers.PretrainedConfig) -> transformers.PreTrained
...
+    fqn: str,  # stands for fully qualified name
+    parent_module: Optional[torch.nn.Module] = None,
+):
```
Comment (Priya2698) [inline tests/python/multidevice/test_deepseek_v3.py:122]: Can we code comment what this function is doing: Recursively find all the linear modules, and parallelize them with nvfuser if a parallel_plan is found

File: tests/python/multidevice/test_deepseek_v3.py [hunk 6]
```diff
@@ -120,19 +192,25 @@ def test_transformer_layer(setup_default_process_group):
...
+                distributed_params: list[str] = [
+                    name
+                    for name, parameter in transformer_layer.named_parameters()
+                    if isinstance(parameter.data, DTensor)
+                ]
+                assert len(distributed_params) == 3 + (config.n_routed_experts + 1) * 3
+            case Executor.NVFUSER:
+                parallelize_module_with_nvfuser(
+                    transformer_layer, mesh, parallel_plan, ""
+                )
```
Comment (Priya2698) [inline tests/python/multidevice/test_deepseek_v3.py]:
```Suggestion
parallelize_module_with_nvfuser(
                    transformer_layer, mesh, parallel_plan, fqn=""
                )
```

So I don't have to switch between functions to understand what the empty string indicates.

File: tests/python/multidevice/test_deepseek_v3.py [hunk 7]
```diff
@@ -62,12 +69,77 @@ def load_model(config: transformers.PretrainedConfig) -> transformers.PreTrained
...
+
+    assert len(parallel_style.output_layouts) == 1, "Expect 1D mesh"
+    output_layout = parallel_style.output_layouts[0]
+
+    if isinstance(parallel_style, RowwiseParallel):
+        assert input_layout.is_shard(-1), "We only implemented TP."
+        assert output_layout.is_replicate(), "We only implemented TP."
+        return TensorParallelLinear.distribute(
+            linear, mesh, [input_layout], [Shard(-1)]
+        )
```
Comment (Priya2698) [inline tests/python/multidevice/test_deepseek_v3.py]:
```Suggestion
        return TensorParallelLinear.distribute(
            linear, mesh, in_placements=[input_layout], weight_placements=[Shard(-1)]
        )
```

Comment (Priya2698) [reply tests/python/multidevice/test_deepseek_v3.py]: Similarly for the next call.

File: tests/python/multidevice/linear.py [hunk 8]
```diff
@@ -30,29 +38,60 @@ def define_linear_forward(config: LinearConfig, fd: FusionDefinition) -> None:
...
         ctx,
         input: DTensor,
         weight: DTensor,
     ):
-        op = FusionDefinitionWrapper(
-            partial(define_linear_forward, LinearConfig(weight.size(1), weight.size(0)))
+        assert input.dim() in [2, 3]
+        op = get_fusion_definition_wrapper(
+            ComputeType.FORWARD,
+            LinearConfig(weight.size(1), weight.size(0), input.dim() == 3),
```
Comment (Priya2698) [inline tests/python/multidevice/linear.py]:
```Suggestion
            LinearConfig(weight.size(1), weight.size(0), has_batch = input.dim() == 3)

File: tests/python/multidevice/linear.py [hunk 9]
```diff
@@ -30,29 +38,60 @@ def define_linear_forward(config: LinearConfig, fd: FusionDefinition) -> None:
...
+    grad = fd.define_tensor(grad_shape, contiguity=True, dtype=DataType.BFloat16)
 
     grad_x = fd.ops.matmul(grad, w)
 
-    grad_flat_t = fd.ops.permute(fd.ops.reshape(grad, [-1, e_out]), [1, 0])
-    x_flat = fd.ops.reshape(x, [-1, e_in])
+    grad_flat_t = grad
+    if config.has_batch:
+        grad_flat_t = fd.ops.reshape(grad_flat_t, [-1, e_out])
+    grad_flat_t = fd.ops.permute(grad_flat_t, [1, 0])
```
Comment (Priya2698) [inline tests/python/multidevice/linear.py]: nit:
```Suggestion
    grad_flat = fd.ops.reshape(grad, [-1, e_out]) if config.has_batch else grad
    grad_flat_t = fd.ops.permute(grad_flat_t, [1, 0])
```
Similarly for `x` below.

File: tests/python/multidevice/linear.py [hunk 10]
```diff
@@ -61,11 +100,47 @@ def forward(
...
+    @classmethod
+    def distribute(
+        cls,
+        linear: torch.nn.Linear,
+        mesh: dist.device_mesh.DeviceMesh,
+        in_placements: Iterable[Placement] = [],
+        weight_placements: Iterable[Placement] = [],
+    ):
+        tp_linear = cls(linear.in_features, linear.out_features, in_placements)
+        tp_linear.weight = torch.nn.Parameter(
```
Comment (Priya2698) [inline tests/python/multidevice/linear.py:137]: `weights` will be a fusion input and hence, given to us.
So should we expect weights to be provided in the initialization of this method, and check their shardings in the `parallelize_linear_with_nvfuser` module?

Comment (wujingyue) [reply tests/python/multidevice/linear.py:137]: I added some code comments in __init__ that hopefully answer this question.

File: tests/python/multidevice/test_deepseek_v3.py (hunk 3 above)
Comment (wujingyue) [reply tests/python/multidevice/test_deepseek_v3.py]: "which expects outputs to be replicated" doesn't add information `output_layout.is_replicate`.

File: tests/python/multidevice/test_deepseek_v3.py (hunk 4 above)
Comment (wujingyue) [reply tests/python/multidevice/test_deepseek_v3.py:119]: What do you mean by this?

File: tests/python/multidevice/test_deepseek_v3.py (hunk 3 above)
Comment (wujingyue) [reply tests/python/multidevice/test_deepseek_v3.py]: Printed output_layout to make the message more useful.

**Comment by wujingyue** [comment]
!test

**Comment by wujingyue** [comment]
!test

File: tests/python/multidevice/test_deepseek_v3.py (hunk 4 above)
Comment (Priya2698) [reply tests/python/multidevice/test_deepseek_v3.py:119]: Sorry about the vague request.
I meant specifying the available parallel styles, but I see now it is a standard import from `torch.dist` so it is not necessary. I am not very familiar with the module, so I did not know what the values would be.

**Comment by wujingyue** [comment]
!test

**Comment by wujingyue** [comment]
!test
